deeppavlov/dataset_iterators/morphotagger_iterator.py

```diff
@@ -103,14 +103,15 @@ def gen_batches(self, batch_size: int, data_type: str = 'train',
         if shuffle is None:
             shuffle = self.shuffle
         data = self.data[data_type]
-        if shuffle:
-            random.shuffle(data)
         lengths = [len(x[0]) for x in data]
         indexes = np.argsort(lengths)
         L = len(data)
         if batch_size < 0:
             batch_size = L
-        for start in range(0, L, batch_size):
+        starts = list(range(0, L, batch_size))
+        if shuffle:
+            np.random.shuffle(starts)
+        for start in starts:
             indexes_to_yield = indexes[start:start + batch_size]
```

Comment:
Do you really need the `indexes` variable?
```suggestion
            indexes_to_yield = list(range(start, start + batch_size))
```

Reply:
Yes, I need them in order to use this generator for inference (I should know the indexes of the elements in current batch).